File: leetcode/Bit_Manipulation/0067_Add_Binary.cpp
class Solution {
public:
    string addBinary(string a, string b) {
        int flagA = a.length() - 1, flagB = b.length() - 1, carry = 0;
        string result;
        while (flagA >= 0 && flagB >= 0) {
        	int temp = (a[flagA] - 48) + (b[flagB] - 48) + carry;
        	carry = temp / 2;
        	result.insert(result.begin(), (temp % 2) + 48);
        	flagA --;
        	flagB --;
        }
        while (flagA >= 0) {
        	int temp = (a[flagA] - 48) + carry;
        	carry = temp / 2;
        	result.insert(result.begin(), (temp % 2) + 48);
        	flagA --;
        }
        while (flagB >= 0) {
			int temp = (b[flagB] - 48) + carry;
        	carry = temp / 2;
        	result.insert(result.begin(), (temp % 2) + 48);
        	flagB --;
        }
        if (carry != 0)
        	result.insert(result.begin(), carry + 48);

        return result;
    }
};

File: leetcode/Bit_Manipulation/0067_Add_Binary.py
class Solution(object):
    def addBinary(self, a, b):
        """
        :type a: str
        :type b: str
        :rtype: str
        """
        len_a = len(a)
        len_b = len(b)
        # convert str to int list, least significant bit on the left
        a_bits = [int(i) for i in a[::-1]]
        b_bits = [int(i) for i in b[::-1]]
        overflow = 0
        result = []

        for i in range(max(len_a, len_b)):
            if i >= len_a:
                bit_sum = b_bits[i] + overflow
            elif i >= len_b:
                bit_sum = a_bits[i] + overflow
            else:
                bit_sum = a_bits[i] + b_bits[i] + overflow

            result.append(bit_sum % 2)
            overflow = bit_sum // 2

        if overflow:
            result.append(overflow)

        result = [str(bit) for bit in result]
        result = ''.join(result)[::-1]
        return result
